nvme_gw_server.py

```diff
@@ -86,57 +183,96 @@ def start_spdk(self):
             raise
         return
 
-    def bdev_rbd_create(self, request, context):
-        # Create bdev from a given RBD image
+    def restore_config(self):
+        callbacks = {}
+        callbacks[self.persistent_config.BDEV_PREFIX] = self.bdev_rbd_create
+        callbacks[self.persistent_config.
+                  SUBSYSTEM_PREFIX] = self.nvmf_create_subsystem
+        callbacks[self.persistent_config.
+                  NAMESPACE_PREFIX] = self.nvmf_subsystem_add_ns
+        callbacks[
+            self.persistent_config.HOST_PREFIX] = self.nvmf_subsystem_add_host
+        callbacks[self.persistent_config.
+                  ALLOW_ALL_PREFIX] = self.nvmf_subsystem_allow_any_host
+        callbacks[self.persistent_config.
+                  TRANSPORT_PREFIX] = self.nvmf_create_transport
+        callbacks[self.persistent_config.
+                  LISTENER_PREFIX] = self.nvmf_subsystem_add_listener
+        self.persistent_config.restore(callbacks)
+
+    def bdev_rbd_create(self, request, context=None):
+        """Creates bdev from a given RBD image."""
         self.logger.info({
-            f"Received: {request.ceph_pool_name}, {request.rbd_name}, {request.block_size}",
+            f"Received request to create bdev {request.bdev_name} from" +
+            f" {request.ceph_pool_name}/{request.rbd_name}",
         })
         try:
             bdev_name = self.spdk_rpc.bdev.bdev_rbd_create(
-                self.client,
-                request.ceph_pool_name,
-                request.rbd_name,
-                request.block_size,
+                self.spdk_rpc_client,
+                name=request.bdev_name,
+                pool_name=request.ceph_pool_name,
+                rbd_name=request.rbd_name,
+                block_size=int(request.block_size),
             )
             self.logger.info(f"Created bdev {bdev_name}")
 
         except Exception as ex:
             self.logger.error(f"bdev create failed with: \n {ex}")
-            context.set_code(grpc.StatusCode.INTERNAL)
-            context.set_details(f"{ex}")
+            if context:
+                context.set_code(grpc.StatusCode.INTERNAL)
+                context.set_details(f"{ex}")
             return pb2.bdev_info()
 
+        if context:
+            # Update persistent configuration
+            try:
+                self.persistent_config.add_bdev(bdev_name, str(request))
+            except Exception:
+                # Stop server manually - cannot raise exception due to gRPC thread
+                # hanging during server deallocation
+                self.server.stop(None)
+                sys.exit(1)
```

Comment:
Shouldn't we print a meaningful error message before terminating? I would wrap this into some `terminate(msg)` function, that would print the message to the log, stop the server and exit.

Alternatively, do we want to terminate the gateway on any error? It might just log the error and proceed, serving other requests. This might be improved in further PRs though.